File: leetcode/problems/problem-specific/104.maximum-depth-of-binary-tree/104.maximum-depth-of-binary-tree.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	return findMax(maxDepth(root.Left), maxDepth(root.Right)) + 1
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/sessions/2023-05-01_2023-??-??/104.maximum-depth-of-binary-tree.py
#
# @lc app=leetcode id=104 lang=python3
#
# [104] Maximum Depth of Binary Tree
#

# @lc code=start

# * Recursive DFS Solution | O(n) Time | O(h) Space
# * n -> The number of nodes in the tree | h -> The height of the tree


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        left_node_depth = self.maxDepth(root.left)
        right_node_depth = self.maxDepth(root.right)
        return 1 + max(left_node_depth, right_node_depth)


# @lc code=end
